### ISSUE TITLE
Save the guids in a lowercase format when syncing sql server with sqlite

### ISSUE THREAD

**Issue by hakakou** [Mimetis/Dotmim.Sync#1069]
Using the latest version of Dotmim.Sync, I'm syncing an SQL Server with an Sqlite.

My problem is that Guid columns in sql server, are converted to an uppercase format in the CHAR(36) of sqlite.

Is there any way to have them written as lowercase guids? The ORM I am using seems to get confused when mapping Guids to the native string, as it uses lowercase in it's select/update/insert statements, and so it cannot match the Uppercase guids written by Dotmim.Sync.

**Comment by workgroupengineering** [comment]
Have you tried to implement a custom converter like [this](https://github.com/Mimetis/Dotmim.Sync/blob/b1d7fa923b5aab677fb29298993a2836d1d313a1/Samples/Dotmim.Sync.SampleConsole/CustomConverter.cs)?

**Comment by hakakou** [comment]
Thanks for the tip! I will try it out.

**Comment by gaurakshay** [comment]
@hakakou Sorry to bother you. But we're you able to implement a converter that would convert the guids to lowercase on sync?

**Comment by Mimetis** [comment]
What the interest ?

**Comment by gaurakshay** [comment]
@Mimetis  of course.

### Background

I am using the `Dotmim.Sync` library (thank you for your amazing work!) to sync data to and from our clients which are `Android` and `iOS` only and it works perfectly. For local database management on these mobile devices, I use `sqlite-net` library (using the `sqlcipher` version, if that makes any difference). 

Please note that all the tables in the database use GUID as the primary key which is stored as byte in the SQLServer and as string in the client's sqlite database.

### Observations
When data is synced _from_ the `server` _to_ a mobile device, the device is not able to perform any updates to the rows that it obtained from the `server`.  The only rows it is able to update are the ones it generates locally. 

The reason for me to believe this is that I am able to run a query using `sqlite-net` method `conn.Table<TEntity>().ToList().Count` that returns 36 (number of rows is not important), but when taking one of the rows and using `conn.Update(TEntity)` to update, the result number of rows updated is always returned as 0.
However, if I add any new rows locally using the `sqlite-net` library, I am always able to update them successfully. I have been trying to figure out any possible reason for such an odd behaviour. I was able to see that all the records synced from the server (not generated on the device) had their primary keys in the uppercase while all the records generated locally had the primary key values in lowercase.

Since `sqlite-net`'s update method requires a primary key to match the object to update, it is my hunch that the mismatch of the upper-case primary key and the lowercase primary key is causing the updates to fail.

### Attempts

I have attempted to write a basic converter using the samples provided in the documentation. and written this:

```csharp
public class GuidCaseConverter : IConverter
{
    public string Key => "cuscom";

    public void BeforeSerialize(SyncRow row)
    {
        // Convert all Guid columns to lowecase
        foreach (var col in row.Table.Columns.Where(c => c.GetDataType() == typeof(Guid)))
        {
            if (row[col.ColumnName] != null)
                row[col.ColumnName] = row[col.ColumnName].ToString().ToLowercaseInvariant();
        }
    }

    public void AfterDeserialized(SyncRow row)
    {
        // Convert all Guid columns to lowecase
        foreach (var col in row.Table.Columns.Where(c => c.GetDataType() == typeof(Guid)))
        {
            if (row[col.ColumnName] != null)
                row[col.ColumnName] = row[col.ColumnName].ToString().ToLowercaseInvariant();
        }
    }
}
```

After adding the converter, the sync happens successfully, however, I think I am making a mistake somewhere because the local sqlite database that I get still has the primary keys in the uppercase.

### Notes

This issue not recreatable on my windows machine.

**Comment by hakakou** [comment]
In my case I solved the problem by modifying the ORM and not dotmim.sync: There was a setting for it to map guids to uppercase.

**Comment by Mimetis** [comment]
> In my case I solved the problem by modifying the ORM and not dotmim.sync: There was a setting for it to map guids to uppercase.

Can you share your code @hakakou  ?

**Comment by Mimetis** [comment]
Looking at the code in the `SQLITE` provider, every Primary Key (especially `GUID`) are created with the `COLLATE NOCASE` property.

`Sqlite/SqliteTableBuilder.cs` :

``` cs
//check if it's a primary key, then, even if it's case sensitive, we turn on case insensitive
if (SyncGlobalization.IsCaseSensitive())
{
    if (this.TableDescription.PrimaryKeys.Contains(column.ColumnName))
        casesensitive = "COLLATE NOCASE";
}
```

What it means is that the column is case insensitive.

Here is the table script generated, as example:

``` cs
CREATE TABLE IF NOT EXISTS [ProductCategory] (
	[ProductCategoryID] text  NOT NULL COLLATE NOCASE
	,[ParentProductCategoryID] text  NULL COLLATE NOCASE
	,[Name] text  NOT NULL COLLATE NOCASE
	,[rowguid] text  NULL COLLATE NOCASE
	,[ModifiedDate] text  NULL 
	,[IsActive] integer  NOT NULL 
	,PRIMARY KEY ([ProductCategoryID])
	FOREIGN KEY ( [ParentProductCategoryID]) REFERENCES [ProductCategory]( [ProductCategoryID] )
)
```
As you can see the `ProductCategoryID` column is generated with a `COLLATE NOCASE` property (and also all the others as the `SyncGlobalization` static class is specifying `DataSourceStringComparison` to `StringComparison.InvariantCultureIgnoreCase`)

**Comment by hakakou** [comment]
> > In my case I solved the problem by modifying the ORM and not dotmim.sync: There was a setting for it to map guids to uppercase.
> 
> Can you share your code @hakakou ?

Yes, but it's very specific to XPO ORM which I used: https://www.devexpress.com/products/net/orm/

```cs
using System;
using System.Data;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.DB.Helpers;

namespace Haka
{
    public class SQLiteConnectionProvider2 : SQLiteConnectionProvider
    {
        public new const string XpoProviderTypeString = "SQLite2";

        public new static IDataStore CreateProviderFromString(string connectionString, AutoCreateOption autoCreateOption, out IDisposable[] objectsToDisposeOnDisconnect)
        {
            IDbConnection dbConnection = CreateConnection(connectionString);
            objectsToDisposeOnDisconnect = new IDisposable[1] { dbConnection };
            return CreateProviderFromConnection(dbConnection, autoCreateOption);
        }

        public new static IDataStore CreateProviderFromConnection(IDbConnection connection, AutoCreateOption autoCreateOption)
        {
            return new SQLiteConnectionProvider2(connection, autoCreateOption);
        }

        static SQLiteConnectionProvider2()
        {
            RegisterDataStoreProvider(XpoProviderTypeString, new DataStoreCreationFromStringDelegate(CreateProviderFromString));
            RegisterFactory(new SQLiteProviderFactory2());
        }

        public SQLiteConnectionProvider2(IDbConnection connection, AutoCreateOption autoCreateOption) : base(connection, autoCreateOption)
        {
        }

        public static new string GetConnectionString(string database)
        {
            return string.Format("{1}={2};Data Source={0}", ConnectionProviderSql.EscapeConnectionStringArgument(database), "XpoProvider", "SQLite2");
        }

        public static new string GetConnectionString(string database, string password)
        {
            return string.Format("{2}={3};Data Source={0};Password={1}", ConnectionProviderSql.EscapeConnectionStringArgument(database), ConnectionProviderSql.EscapeConnectionStringArgument(password), "XpoProvider", "SQLite2");
        }

        public new static void Register() { }

        protected override object ConvertToDbParameter(object clientValue, TypeCode clientValueTypeCode)
        {
            if (clientValueTypeCode == TypeCode.Object && clientValue is Guid)
            {
                return ((Guid)clientValue).ToString().ToUpper();
            }
            return base.ConvertToDbParameter(clientValue, clientValueTypeCode);
        }
    }

    public class SQLiteProviderFactory2 : SQLiteProviderFactory
    {
        public override string ProviderKey => "SQLite2";
    }

}
```

**Comment by Mimetis** [comment]
I tried **sqlite-net** to see what is the problem.

First of all, I've used **DMS** to sync the `ProductCategory` table (from **SQL Server**) to a **SQLITE** database.
As expected, the primary key is stored in **Upper Case**:

![image](https://github.com/Mimetis/Dotmim.Sync/assets/4592555/681886ef-012d-48e3-9154-df807311ec90)

Using **sqlite-net**, I've made some samples queries, mixing upper case / lower case

``` cs
var lowerCase = "0a609694-c6a6-4541-9385-0b7a877a7317".ToLower();
var upperCase = "0a609694-c6a6-4541-9385-0b7a877a7317".ToUpper();

var id1 = new Guid(lowerCase);
var item = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id1);

Console.WriteLine($"LowerCase: {item.ProductCategoryID.ToString()}: {item.Name}");

var id2 = new Guid(upperCase);
var item2 = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id2);

Console.WriteLine($"UpperCase: {item2.ProductCategoryID.ToString()}: {item2.Name}");
```
And it's working as expected:

``` bash
LowerCase: 0a609694-c6a6-4541-9385-0b7a877a7317: Shoes
UpperCase: 0a609694-c6a6-4541-9385-0b7a877a7317: Shoes
```
I've also made some tests using **Update** queries:

``` cs
var lowerCase = "0a609694-c6a6-4541-9385-0b7a877a7317".ToLower();
var upperCase = "0a609694-c6a6-4541-9385-0b7a877a7317".ToUpper();

var id1 = new Guid(lowerCase);
var item1 = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id1);

item1.Name = "Shoes Updated To Lower Case";
db.Update(item1);

// query again, to be sure
item1 = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id1);

Console.WriteLine($"LowerCase: {item1.ProductCategoryID.ToString()}: {item1.Name}");


var id2 = new Guid(upperCase);
var item2 = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id2);

item2.Name = "Shoes Updated To Upper Case";
db.Update(item2);

// query again, to be sure
item2 = db.Table<ProductCategory>()
    .FirstOrDefault(p => p.ProductCategoryID == id2);

Console.WriteLine($"UpperCase: {item2.ProductCategoryID.ToString()}: {item2.Name}");
```

And once again, results seems good enough:

``` bash
LowerCase: 0a609694-c6a6-4541-9385-0b7a877a7317: Shoes Updated To Lower Case
UpperCase: 0a609694-c6a6-4541-9385-0b7a877a7317: Shoes Updated To Upper Case
```

Once again, I'm not convinced you need to handle the lower case / upper case collation
Let me know

**Comment by hakakou** [comment]
Obviously in my case the problem is that the ORM creates the table itself, and does not use `COLLATE NOCASE` to allow case-insensitive queries on the string id column.

**Comment by gaurakshay** [comment]
@Mimetis Thank you. I didn't realize that the sync framework uses "COLLATE NOCASE" option. Please give me some to analyze this and whip up a reproducible sample for you if possible.
I'd also like to ask you if it is okay replying on this closed issue. Perhaps it would better if I started a discussion on this topic?

**Comment by gaurakshay** [comment]
So I reviewed the logic at my end to figure out what was wrong. We are using `sqlite-net`'s inbuilt capabilities to create any tables as it automatically creates any extra columns if needed for schema updates. This was resulting in the tables structures not having `COLLATE NOCASE` defined.
To work around this issue, I added the very helpful `COLLATE` attribute on all the GUID properties in my models so that when `sqlite-net` creates any tables, it creates them with the `COLLATE NOCASE` turned on, fixing any issues that I had in my app.

For example:
<img width="605" alt="Screenshot 2024-04-15 at 16 27 52" src="https://github.com/Mimetis/Dotmim.Sync/assets/17280428/631c57b4-934e-4b7f-a134-81f7dd04b087">

Thank you @Mimetis and @hakakou !
